### ISSUE TITLE
Support for running arbitrary main classes

### ISSUE THREAD

**Issue by hmemcpy** [bazelbuild/rules_scala#468]
This is a proposal outlined in a feature request by @ittaiz: https://github.com/bazelbuild/intellij/issues/247

We want to support running arbitrary `main` methods (useful for testing/otherwise running arbitrary code that's not supposed to run outside the IDE). For this, we want to create an implicit output binary target for every `scala_library`, and call it explicitly, passing additional context from the IDE (specifically, the `main_class` value).

Perhaps this could be a joint effort to support @johnynek's [idea for the repl](https://github.com/bazelbuild/rules_scala/issues/448)?

WDYT?

**Comment by johnynek** [comment]
Shouldn’t be too hard once we solve the runfiles issue I had with the repl.

So, in this case we have a wrapper script that expects a classname as a first argument and the rest are passed into main?

It would be interesting to ask the bazel people why they didn’t do this with java_library, or if they did how to copy it so it works exactly the same way.

**Comment by ittaiz** [comment]
From this [comment](https://github.com/bazelbuild/intellij/issues/270#issuecomment-376299483) it sounds this doesn't exist in java:
 

> ...Bazel doesn't build a runnable binary for java_library targets.

**Comment by ittaiz** [comment]
@hmemcpy can you please update here about the difficulties and what we decided to do?

**Comment by ittaiz** [comment]
@hmemcpy
Can you update about resolution and optionally close this

**Comment by hmemcpy** [comment]
Oops, sorry, must have missed this!
We decided to implement this entirely in the plugin, without any additional changes in rules_scala, by creating a deployable.jar out of the module where the arbitrary main was added, and then running it with the `-cp` argument, passing it the fully qualified name of the `main` object. This supports both run and debug scenarios, allowing to quickly run any `scala_library` without any changes to BUILD files or other local changes!
I'm closing this issue, as well as the issue in the plugin repo!
